rfd/0101-pod-rbac.md

```diff
@@ -0,0 +1,825 @@
+handle them separately.
+
+##### Listing Pods
+
+When listing Pods in the cluster or at a certain namespace, Teleport will receive
+the list of Pods from Kubernetes itself. Since Teleport depends on what Kubernetes
+RBAC allows users to access, Teleport would need to hijack the Kubernetes API
+response to exclude any Pod that is not included in the allowed list.
+
+To do that, Teleport will automatically forward the request to Kubernetes
+API for the following two endpoints:
+
+- `/api/v1/pods`
+  - `GET`: list or watch objects of kind Pods in every namespace.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `GET`: list or watch objects of kind Pods in namespace `{namespace}`.
+
+Once Teleport receives the response from Kubernetes API, it will unmarshal the
+response and exclude any Pod that is not in the allowed list. After excluding them,
+Teleport will return the filtered response to the user.
+
+##### Deleting Pods
+
+Kubernetes allows a user to delete every Pod in a certain namespace using a single
+call. The call does not include the name of the Pods to delete, so Teleport won't
+forward the request to Kubernetes. Instead, Teleport will list the user's Pods,
+exclude the ones that he shouldn't have access to, and issue a delete request for
+each one of the resulting Pod names.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `DELETE`: delete a collection of Pods in namespace `{namespace}`.
+
+Another option is to prevent this type of request in Teleport. This means that
+if you try to delete every Pod in the namespace while your role limits the Pods
+you have access to, Teleport will reject the request and inform the user that
+he is not allowed to do it. This option will restrict one of the `kubectl`
+functionalities.
+
+##### Creating Pods
+
+The endpoint for creating pods won't be restricted, and if restrictions exist
+it's the Kubernetes RBAC layer that applies them.
+
+- `/api/v1/namespaces/{namespace}/pods`
+  - `POST`: create a POD in namespace `{namespace}`.
+
+### Security
+
+When a user has multiple roles assigned and his access is expected to be restricted
+to a subset of the pods available, it's a requirement that the union of every
+`kubernetes_resources` defined does not give broader permissions as exemplified by `user4` in
+[Restricted Pod Access](#restrict-pod-access-selected) examples. Otherwise,
+Teleport will leak the Pods when the user lists them.
+
+When a user lists pods - `kubectl get pods`, using the Kubernetes RBAC given by
+`search_as_roles` or by a Kubernetes RBAC principal that allows listing pods
+before requesting access to a certain Pod, the user has access
+to other details besides the Pod name. It happens because `kubectl get pods -o json|yaml`
+API response returns Pod description that includes env variables, secrets used,
+and other details. This might be a security concern because Pods YAML might contain
+secrets that can be leaked even before creating an access request.
```

Comment:
Is there anything that can be done to mitigate the risk of leaking env vars and secrets of pods that a given Teleport user doesn't have access to? Such leak feels pretty compromising as it might lead to an entire bypass of the access and visibility limitations that the Teleport is trying to enforce.

At the very least is there a possibility to detect such cases and warn admins when the configuration allows for such bypasses?

Reply:
I didn't implement this part of the RFD because I have the same concerns.
For now, I implemented a `tsh request search --kind=pod` that does the pod `LIST` request in Teleport AuthServer and returns the `name`, `labels`, and the `namespace` for each pod - not yet merged into master (#19791). We are still discussing if it's the correct approach since it requires Auth Server to connect to a Teleport Proxy. 

With `tsh request search --kind=pod`, we avoid leaking env's and other critical information that, generally, the user must not have access to without requesting access.

Reply:
@tigrato 
Does implementing the `tsh request search --kind=pod` really prevent the user from running `kubectl get pods -o json|yaml` with the same privs and accessing the dangerous information anyway?

Reply:
@SzymonDrosdzol 
Yes.

If you do `kubectl get pods` you just have access to the pods that your role or, in case of an access request, your access request specifies. No other pods will be included.

`tsh request search --kind=pod` will allow you to see what is available in the cluster but will return the pod `name`, `namespace`, and its `labels` hiding every other pod item that might contain secrets or secret refs.

Reply:
Got it. Sounds good @tigrato